### DISCUSSION TITLE
How can I prevent user creation without authentication provider

### DISCUSSION THREAD

**Discussion by Phaiax** [pocketbase/pocketbase#3488] [Q&A] [answered]
Hello

I'm building a sveltekit+pocketbase app. 

For user authentication, I followed this guide: https://dev.to/mpiorowski/oauth2-is-so-complicated-or-90-lines-of-code-with-svelte-532

I want my users to be authenticated only via authentication providers (oauth2), and want to forbid registering/authentication via email+password.

I can't come up with a `create User` API rule that forbids this. Setting the API rule it to `admins only` does not seem to work because `await pb.collection('users').authWithOAuth2({ provider: 'github' });` which is executed on the client, already creates a record. The API filter does not allow to filter like `password == ""` field.

I also can't prevent logging in, because the backend is supposed to be accessible by the client, circumventing any logic I could have in my sevelte app.

The only way I can think of would be to check in every relevant API rule of all collections, that the user is `verified` and sometimes purge non-verified users.

Best regards,
Daniel

This is how I 'hack' my app:

```js
var script = document.createElement('script');
script.src = 'node_modules/pocketbase/dist/pocketbase.umd.js';
document.head.appendChild(script);
//
pb = new PocketBase('http://localhost:8090')
await pb.collection('users').create({username: "user6", password: "123456789", passwordConfirm:"123456789", verified: false, name:"", email:"abc@def.ghi"})
// -> Object { avatar: "", collectionId: "_pb_users_auth_", collectionName: "users", created: "2023-10-10 16:02:58.002Z", emailVisibility: false, id: "dta7t1mapieuhyo", name: "", updated: "2023-10-10 16:02:58.002Z", username: "user6", verified: false }
pb.authStore.isValid
// -> false
const userData = await pb.collection('users').authWithPassword('abc@def.ghi', '123456789');
// -> Object { record: {…}, token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJfcGJfdXNlcnNfYXV0aF8iLCJleHAiOjE2OTgxNjM1MDQsImlkIjoiZHRhN3QxbWFwaWV1aHlvIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.X0xDvyqCrUVT9HKIdus3Bj0kFgjWhVPl7M5GMhOZ3ws" }
pb.authStore.isValid
// -> true
await pb.collection('users').getFullList();
// -> Array [ {…} ]
//    0: Object { collectionId: "_pb_users_auth_", collectionName: "users", created: "2023-10-10 16:02:58.002Z", … }
await pb.collection('users').update('dta7t1mapieuhyo', {name: "Hi Mom"});
// -> Object { avatar: "", collectionId: "_pb_users_auth_", collectionName: "users", created: "2023-10-10 16:02:58.002Z", email: "abc@def.ghi", emailVisibility: false, id: "dta7t1mapieuhyo", name: "Hi Mom", updated: "2023-10-10 16:09:29.310Z", username: "user6", … }
```

​

**Comment by ganigeorgiev** [comment, marked as answer]
If you just want to disable the password login, you can go to:
_Admin UI > Collections > {your_auth_collection} > Edit collection (settings cogwheel) > Options_ and disable the toggles for `Username/Password` and `Email/Password`.

**Comment by Phaiax** [comment]
Thanks for the quick reply, @ganigeorgiev  😄 . I overlooked these options.

With these settings, password authentication fails, but everybody is still able to create user records.

But I figured out how to prevent record creation after writing the question 😅 :

Setting `verified = true` in the `create Rule` of `users` prevents creating the user through the API. (Pocketbase itself has special handling for this field and does only allow admins to set it to true). The initial login aka registering via Oauth2 still works. It seems that there, the user record is created with admin credentials. (But I wonder why oauth2 then didn't work with `admin-only`)

**Reply by ganigeorgiev** [reply]
> But I wonder why oauth2 then didn't work with admin-only

The OAuth2 created user must satisfy the Create API rule of the collection.
